src/Voron/Impl/Journal/JournalReader.cs

```diff
@@ -245,7 +245,20 @@ public bool ReadOneTransactionToDataFile(ref Pager.State dataPagerState, ref Pag
 
             if ((current->Flags & TransactionPersistenceModeFlags.HasFreePages) == TransactionPersistenceModeFlags.HasFreePages)
             {
-                var freePages = ReadEncodedFreePage(outputPage + totalRead, _allocator);
+                var freePagesPtr = outputPage + totalRead;
+
+                var header = *(FreePagesHeader*)freePagesPtr;
+                
+                var freePagesSpaceToRead = sizeof(FreePagesHeader) + header.EncodedSectionsSize + header.EncodedSectionsCount * sizeof(EncodedFreePagesSection);
+
+                if (totalRead + freePagesSpaceToRead > current->UncompressedSize)
+                {
+                    throw new InvalidDataException(
+                        $"Attempted to read free pages from position {totalRead} to {totalRead + freePagesSpaceToRead} (size: {freePagesSpaceToRead} bytes) from transaction data while " +
+                        $"the transaction uncompressed size is {current->UncompressedSize} bytes");
+                }
+
+                var freePages = ReadEncodedFreePage(freePagesPtr, _allocator);
```

Comment:
I think we need to pass `UncompressedSize` here as well.
If we have a bad value in one of the segments size, we need to error there as well